codec-redis/src/main/java/io/netty/handler/codec/redis/RedisEncoder.java

```diff
@@ -0,0 +1,154 @@
+            writeBulkStringHeader(allocator, (BulkStringHeaderRedisMessage) msg, out);
+        } else if (msg instanceof ArrayHeaderRedisMessage) {
+            writeArrayHeader(allocator, (ArrayHeaderRedisMessage) msg, out);
+        } else if (msg instanceof ArrayRedisMessage) {
+            writeArrayMessage(allocator, (ArrayRedisMessage) msg, out);
+        } else {
+            throw new CodecException("bad message: " + msg);
+        }
+    }
+
+    private static void writeInlineByteBufMessage(ByteBufAllocator allocator, StringRedisMessage msg,
+                                                  List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        out.add(ByteBufUtil.encodeString(allocator, CharBuffer.wrap(msg.content()), CharsetUtil.UTF_8));
+        out.add(EOL.slice());
+    }
+
+    private static void writeIntegerMessage(ByteBufAllocator allocator, IntegerRedisMessage msg, List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        out.add(numberToByteBuf(msg.value()));
+        out.add(EOL.slice());
+    }
+
+    private static void writeBulkStringHeader(ByteBufAllocator allocator, BulkStringHeaderRedisMessage msg,
+                                              List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        if (msg.isNull()) {
+            out.add(NULL_BULK_STRING.slice());
+        } else {
+            out.add(numberToByteBuf(msg.bulkStringLength()));
+            out.add(EOL.slice());
+        }
+    }
+
+    private static void writeBulkStringContent(BulkStringRedisContent msg, List<Object> out) {
+        out.add(msg.content().retain());
+        if (msg instanceof LastBulkStringRedisContent) {
+            out.add(EOL.slice());
+        }
+    }
+
+    private static void writeFullBulkStringMessage(ByteBufAllocator allocator, FullBulkStringRedisMessage msg,
+                                               List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        if (msg.isNull()) {
+            out.add(NULL_BULK_STRING.slice());
+        } else {
+            ByteBuf content = msg.content();
+            out.add(numberToByteBuf(content.readableBytes()));
+            out.add(EOL.slice());
+            out.add(content.retain());
+        }
+        out.add(EOL.slice());
+    }
+
+    /**
+     * Write array header only without body. Use this if you want to write arrays as streaming.
+     */
+    private static void writeArrayHeader(ByteBufAllocator allocator, ArrayHeaderRedisMessage msg, List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        if (msg.isNull()) {
+            out.add(NULL_BULK_STRING.slice());
+        } else {
+            out.add(numberToByteBuf(msg.length()));
+        }
+        out.add(EOL.slice());
```

Comment:
Some of these encoder use cases may be better served by allocating a single buffer, writing to it, and then adding that single buffer to `out`. For example `EOL` represents 2 bytes of data, but creating a slice will create a new object which has a reference to a ByteBuf and 2 other `int`s for offset (not to mention other bounds checking etc...)... which feels like a relatively large overhead to encode 2 bytes.

```java
ByteBuf buffer = allocator.ioBuffer(1 + 20 + 2);
out.add(buffer);
if (is null) {
  buffer.write(...);
} else {
  writeNumberToByteBuf(buf, msg.length());
  buffer.writeShort(EOL_AS_SHORT);
}
```